easypy/caching.py

```diff
@@ -179,12 +180,14 @@ def inner(*args, **kwargs):
                 result, ts = cache.get(key, NOT_CACHED)
 
                 if expiration > 0:
-                    if result is not NOT_FOUND and time.time() - ts > expiration:
+                    if result is not NOT_FOUND and get_ts_func() - ts > expiration:
                         result = NOT_FOUND
                         del cache[key]
 
                 if result is NOT_FOUND:
-                    ts = time.time()
+                    if log_recalculation:
+                        _logger.debug('time cache expired, calculating new value for %s', name)
+                    ts = get_ts_func()
```

Comment:
I'm referring to the case of calling `get_ts_func()` it twice:
```python
if expiration > 0:
    # First call to get_ts_func() to check if we need to run func()
    if result is not NOT_FOUND and get_ts_func() - ts > expiration:
        result = NOT_FOUND
        del cache[key]

if result is NOT_FOUND:
    if log_recalculation:
        _logger.debug('time cache expired, calculating new value for %s', name)
    result = func(*args, **kwargs)
    # Second call to get_ts_func() to determine the cache time
    ts = get_ts_func()
    cache[key] = result, ts
```
The main point of this approach is that we measure the expiration time from *after* `func()` gets called. So, for example, if `expiration=10` and `func()` takes 5 seconds to run, in the current implementation:

- t+0: Call `func()`. Set cache expiration time to `t+10`
- t+5: `func()` returns
- t+10: `func()` expires

So the result expires 5 seconds after we get it. But with the approach I described above:

- t+0: Call `func()`
- t+5: `func()` returns. Set cache expiration time to `t+15`
- t+15: `func()` expires

So the result expires 10 seconds after we get it.